packages/beacon-node/src/chain/stateCache/lruBlockStateCache.ts

```diff
@@ -0,0 +1,165 @@
+import {toHexString} from "@chainsafe/ssz";
+import {RootHex} from "@lodestar/types";
+import {CachedBeaconStateAllForks} from "@lodestar/state-transition";
+import {routes} from "@lodestar/api";
+import {Metrics} from "../../metrics/index.js";
+import {LinkedList} from "../../util/array.js";
+import {MapTracker} from "./mapMetrics.js";
+import {BlockStateCache} from "./types.js";
+
+export type LRUBlockStateCacheOpts = {
+  maxStates: number;
+};
+
+/**
+ * New implementation of BlockStateCache that keeps the most recent n states consistently
+ *  - Maintain a linked list where the head state is always the first item in the list
+ *  - Prune per add() instead of per checkpoint so it only keeps n historical states consistently, prune from tail
+ *  - This is LRU like cache except that we only track the last added time, not the last used time
+ * because state could be fetched from multiple places, but we only care about the last added time.
+ *  - No need to prune per finalized checkpoint
+ *
+ * Given this block tree with Block 11 as head:
+ *       Block 10
+           |
+     +-----+-----+
+     |           |
+  Block 11   Block 12
+     ^           |
+     |           |
+    head       Block 13
+ *
+ * The maintained key order would be: 11 -> 13 -> 12 -> 10, and state 10 will be pruned first.
+ */
+export class LRUBlockStateCache implements BlockStateCache {
```

Comment:
this is another/new implementation of `BlockStateCache`, same functionality to `StateContextCache`. This will be used in regen, will have feature flag to decide this new state cache or the old `StateContextCache` in the next PR